File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size(),res=0;
        int l=0,r=0;

        map<char,int> map;

        while(r<n){
            if(map.find(s[r])!=map.end() && map[s[r]]>=l){
                l=map[s[r]]+1;
            }
            map[s[r]]=r;
            res=max(r-l+1,res);
            r++;
        }

        return res;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        table = {}
        n=len(s)
        i,j=0,0
        res=0 

        while j!=n:
            table[s[j]] = table.get(s[j],0)+1

            while table[s[j]]>1:
                table[s[i]]=table.get(s[i])-1
                i+=1

            res = max(res,j-i+1)
            j+=1

        return res
